engines/sherlock/scene.cpp

```diff
@@ -0,0 +1,1564 @@
+}
+
+/**
+ * Check the scene's objects against the game flags. If false is passed,
+ * it means the scene has just been loaded. A value of true means that the scene
+ * is in use (ie. not just loaded)
+ */
+void Scene::checkSceneFlags(bool flag) {
+	SpriteType mode = flag ? HIDE_SHAPE : HIDDEN;
+
+	for (uint idx = 0; idx < _bgShapes.size(); ++idx) {
+		Object &o = _bgShapes[idx];
+
+		if (o._requiredFlag) {
+			if (!_vm->readFlags(_bgShapes[idx]._requiredFlag)) {
+				// Kill object
+				if (o._type != HIDDEN && o._type != INVALID) {
+					if (o._images == nullptr || o._images->size() == 0)
+						// No shape to erase, so flag as hidden
+						o._type = HIDDEN;
+					else
+						// Flag it as needing to be hidden after first erasing it
+						o._type = mode;
+				}
+			} else if (_bgShapes[idx]._requiredFlag > 0) {
+				// Restore object
+				if (o._images == nullptr || o._images->size() == 0)
+					o._type = NO_SHAPE;
+				else
+					o._type = ACTIVE_BG_SHAPE;
+			}
+		}
+	}
+
+	// Check inventory for items to remove based on flag changes
+	for (int idx = 0; idx < _vm->_inventory->_holdings; ++idx) {
+		InventoryItem &ii = (*_vm->_inventory)[idx];
+		if (ii._requiredFlag && !_vm->readFlags(ii._requiredFlag)) {
+			// Kill object: move it after the active holdings
+			InventoryItem tempItem = (*_vm->_inventory)[idx];
+			_vm->_inventory->insert_at(_vm->_inventory->_holdings, tempItem);
+			_vm->_inventory->remove_at(idx);
+			_vm->_inventory->_holdings--;
+		}
+	}
+
+	// Check inactive inventory items for ones to reactivate based on flag changes
+	for (uint idx = _vm->_inventory->_holdings; idx < _vm->_inventory->size(); ++idx) {
+		InventoryItem &ii = (*_vm->_inventory)[idx];
+		if (ii._requiredFlag && _vm->readFlags(ii._requiredFlag)) {
+			// Restore object: move it after the active holdings
+			InventoryItem tempItem = (*_vm->_inventory)[idx];
+			_vm->_inventory->remove_at(idx);
+			_vm->_inventory->insert_at(_vm->_inventory->_holdings, tempItem);
+			_vm->_inventory->_holdings++;
+		}
+	}
+}
+
+/**
+ * Checks scene objects against the player's inventory items. If there are any
+ * matching names, it means the given item has already been picked up, and should
+ * be hidden in the scene.
+ */
+void Scene::checkInventory() {
+	for (uint shapeIdx = 0; shapeIdx < _bgShapes.size(); ++shapeIdx) {
+		for (int invIdx = 0; invIdx < _vm->_inventory->_holdings; ++invIdx) {
+			if (_bgShapes[shapeIdx]._name.equalsIgnoreCase((*_vm->_inventory)[invIdx]._name)) {
+				_bgShapes[shapeIdx]._type = INVALID;
+				break;
+			}
+		}
+	}
+}
+
+/**
+ * Set up any entrance co-ordinates or entrance canimations, and then transition
+ * in the scene
+ */
+void Scene::transitionToScene() {
+	People &people = *_vm->_people;
+	SaveManager &saves = *_vm->_saves;
+	Screen &screen = *_vm->_screen;
+	Talk &talk = *_vm->_talk;
+	Common::Point &hSavedPos = people._hSavedPos;
+	int &hSavedFacing = people._hSavedFacing;
+
+	const int FS_TRANS[8] = {
```

Comment:
Might make sense to declare this as static because it seems to be always the same.